<a href="https://colab.research.google.com/github/breadsnicks/AccelerometerClassification/blob/master/ELEC373_Assignment1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import random
from collections import deque

In [181]:
queue = deque()
serviceRate = 0.75
timeslots = 1000000
systemTimes = []
avg_queue_size = 0

In [194]:
# adjust arrival rate
arrivalRate = 0.74

In [183]:
def singleServerQueueSim():
  queue_length_sum = 0
  for t in range(1, timeslots + 1):
    # check if arrival happened -> record the time it arrived calculate time spent in queue
    if random.random() < arrivalRate:
      queue.append(t)

    if queue and random.random() < serviceRate:
      # Record service time and calculate time in queue by determining the difference
      # between time of service and
      arrivalTime = queue.popleft()
      timeInQueue = t - arrivalTime
      systemTimes.append(timeInQueue)

    queue_length_sum += len(queue)

  avg_queue_size = queue_length_sum / timeslots
  avg_time = sum(systemTimes) / len(systemTimes)
  print(f"Final Queue: {list(queue)}")
  print(f"Average Waiting Time:", avg_time)
  print(f"Average Queue Size:", avg_queue_size)
  avg_queuing_delay = avg_queue_size / arrivalRate
  print(f"Queuing Delay via Little's law:",avg_queuing_delay)


In [184]:
singleServerQueueSim()

Final Queue: []
Average Waiting Time: 0.996173193531075
Average Queue Size: 0.497461
Queuing Delay via Little's law: 0.994922



**Results for Each Arrival Rate**:

0.20 Arrival Rate:

> Final Queue: [] <br>
Average Waiting Time: 0.45411668101980734 <br>
Average Queue Size: 0.090698 <br>
Queuing Delay via Little's law: 0.45349

0.40 Arrival Rate:
> Final Queue: [] <br>
Average Waiting Time: 0.7117101173280024 <br>
Average Queue Size: 0.285526 <br>
Queuing Delay via Little's law: 0.713815

0.50 Arrival Rate:
> Final Queue: [999999, 1000000] <br>
Average Waiting Time: 0.9975371577079362 <br>
Average Queue Size: 0.498601 <br>
Queuing Delay via Little's law: 0.997202

0.60 Arrival Rate:
> Final Queue: [] <br>
Average Waiting Time: 1.6841648683930093 <br>
Average Queue Size: 1.011469 <br>
Queuing Delay via Little's law: 1.6857816666666667

0.65 Arrival Rate:
> Final Queue: [] <br>
Average Waiting Time: 2.4641122412713456 <br>
Average Queue Size: 1.601567 <br>
Queuing Delay via Little's law: 2.463949230769231

0.70 Arrival Rate:
> Final Queue: [999994, 999996, 999997, 999998, 999999, 1000000] <br>
Average Waiting Time: 5.115856933105572 <br>
Average Queue Size: 3.581879 <br>
Queuing Delay via Little's law: 5.11697

0.72 Arrival Rate:
> Final Queue: [999986, 999987, 999989, 999990, 999991, 999992, 999993, 999994, 999995, 999996, 999997, 999998, 999999, 1000000] <br>
Average Waiting Time: 8.16097284470662 <br>
Average Queue Size: 5.874261 <br>
Queuing Delay via Little's law: 8.158695833333333

0.74 Arrival Rate:
> Final Queue: [999945, 999946, 999947, 999949, 999950, 999952, 999954, 999955, 999956, 999957, 999958, 999963, 999964, 999965, 999966, 999967, 999968, 999969, 999970, 999971, 999972, 999973, 999974, 999975, 999976, 999977, 999979, 999980, 999981, 999982, 999983, 999985, 999989, 999990, 999991, 999992, 999995, 999996, 999997, 999999, 1000000] <br>
Average Waiting Time: 26.08214362492513 <br>
Average Queue Size: 19.291719 <br>
Queuing Delay via Little's law: 26.06989054054054

0.745 Arrival Rate:
> Final Queue: [999944, 999945, 999947, 999948, 999949, 999950, 999953, 999954, 999955, 999956, 999958, 999959, 999960, 999961, 999962, 999964, 999965, 999966, 999968, 999969, 999970, 999971, 999972, 999973, 999976, 999977, 999979, 999980, 999982, 999983, 999985, 999986, 999990, 999991, 999992, 999994, 999995, 999996, 999997, 999998, 999999, 1000000] <br>
Average Waiting Time: 54.016401602460036 <br>
Average Queue Size: 40.262564 <br>
Queuing Delay via Little's law: 54.04371006711409







In [195]:
# Apply theoretical calcualtions for average waiting time:
theo_wait_time = (1-arrivalRate) / (serviceRate - arrivalRate)
print(theo_wait_time)

25.99999999999998


**Theoretical Wait Times:** <br>
0.20 Arrival Rate
> 1.4545454545454546

0.40 Arrival Rate:
>1.7142857142857144

0.50 Arrival Rate:
> 2.0

0.60 Arrival Rate:
> 2.6666666666666665

0.65 Arrival Rate:
> 3.5000000000000004

0.70 Arrival Rate:
> 5.999999999999996

0.72 Arrival Rate:
> 9.333333333333327

0.74 Arrival Rate:
> 25.99999999999998

0.745 Arrival Rate
> 50.99999999999996